client/app/pages/alert/Alert.jsx

```diff
@@ -0,0 +1,392 @@
+import { TimeAgo } from '@/components/TimeAgo';
+
+import Form from 'antd/lib/form';
+import Button from 'antd/lib/button';
+import Icon from 'antd/lib/icon';
+import Modal from 'antd/lib/modal';
+import Input from 'antd/lib/input';
+import Dropdown from 'antd/lib/dropdown';
+import Menu from 'antd/lib/menu';
+
+import Criteria from './components/Criteria';
+import Rearm from './components/Rearm';
+import Query from './components/Query';
+import { STATE_CLASS } from '../alerts/AlertsList';
+import { routesToAngularRoutes } from '@/lib/utils';
+
+
+const defaultNameBuilder = templateBuilder('<%= query.name %>: <%= options.column %> <%= options.op %> <%= options.value %>');
+const spinnerIcon = <i className="fa fa-spinner fa-pulse m-r-5" />;
+
+function isNewAlert() {
+  return $route.current.params.alertId === 'new';
+}
+
+function HorizontalFormItem({ children, label, className, ...props }) {
+  const labelCol = { span: 4 };
+  const wrapperCol = { span: 16 };
+  if (!label) {
+    wrapperCol.offset = 4;
+  }
+
+  className = cx('alert-form-item', className);
+
+  return (
+    <Form.Item labelCol={labelCol} wrapperCol={wrapperCol} label={label} className={className} {...props}>
+      { children }
+    </Form.Item>
+  );
+}
+
+HorizontalFormItem.propTypes = {
+  children: PropTypes.node,
+  label: PropTypes.string,
+  className: PropTypes.string,
+};
+
+HorizontalFormItem.defaultProps = {
+  children: null,
+  label: null,
+  className: null,
+};
+
+function AlertState({ state, lastTriggered }) {
+  return (
+    <div className="alert-state">
+      <span className={`alert-state-indicator label ${STATE_CLASS[state]}`}>Status: {state}</span>
+      {state === 'unknown' && (
+        <div className="ant-form-explain">
+          Alert condition has not been evaluated.
+        </div>
+      )}
+      {lastTriggered && (
+        <div className="ant-form-explain">
+          Last triggered <span className="alert-last-triggered"><TimeAgo date={lastTriggered} /></span>
+        </div>
+      )}
+    </div>
+  );
+}
+
+AlertState.propTypes = {
+  state: PropTypes.string.isRequired,
+  lastTriggered: PropTypes.string,
+};
+
+AlertState.defaultProps = {
+  lastTriggered: null,
+};
+
+class AlertPage extends React.Component {
+  state = {
+    alert: null,
+    queryResult: null,
+    pendingRearm: null,
+    editMode: false,
+    canEdit: false,
+    saving: false,
+    canceling: false,
+  }
+
+  componentDidMount() {
+    if (isNewAlert()) {
+      this.setState({
+        alert: new AlertService({
+          options: {
+            op: 'greater than',
+            value: 1,
+            rearm: 0,
+          },
+          pendingRearm: 0,
+        }),
+        editMode: true,
+        canEdit: true,
+      });
+    } else {
+      const { alertId } = $route.current.params;
+      const { editMode } = $route.current.locals;
+      AlertService.get({ id: alertId }).$promise.then((alert) => {
```

Comment:
Here and in few places below there is a possible `setState` on unmounted element (which is also possible memleak). We can leave it for now, but should fix it one day

Reply:
I implemented `this._isMounted` in a few places with https://github.com/getredash/redash/pull/4153/commits/52f9ea846650ae78f65a7816f3fa800838c2bf08. Lmk if that's cool.

Reply:
Looks good (hope I didn't miss anything, anyway - it's not so critical). Since I already approved PR - I have nothing to add :+1: